hal/src/common/gpio/v2.rs

```diff
@@ -0,0 +1,1701 @@
+        )+
+    ]) => {
+        paste! {
+            $(
+                $( #[$cfg] )?
+                #[
+                    doc = "Type-level variant of `AlternateConfig` for \
+                    alternate peripheral function " $Letter
+                ]
+                pub enum $Letter {}
+                $( #[$cfg] )?
+                impl Sealed for $Letter {}
+                $( #[$cfg] )?
+                impl AlternateConfig for $Letter {
+                    const CONFIG: AlternateV = AlternateV::$Letter;
+                    const NUM: u8 = $NUM;
+                }
+                $( #[$cfg] )?
+                #[
+                    doc = "Type-level variant of `PinMode` for alternate \
+                    peripheral function " $Letter
+                ]
+                pub type [<Alternate $Letter>] = Alternate<$Letter>;
+            )+
+        }
+    };
+}
+
+alternate!([
+    (A, 00),
+    (B, 01),
+    (C, 02),
+    (D, 03),
+    (E, 04),
+    (F, 05),
+    (G, 06),
+    #[cfg(any(feature = "samd21", feature = "min-samd51g"))]
+    (H, 07),
+    #[cfg(feature = "min-samd51g")]
+    (I, 08),
+    #[cfg(feature = "min-samd51g")]
+    (J, 09),
+    #[cfg(feature = "min-samd51g")]
+    (K, 10),
+    #[cfg(feature = "min-samd51g")]
+    (L, 11),
+    #[cfg(feature = "min-samd51g")]
+    (M, 12),
+    #[cfg(feature = "min-samd51g")]
+    (N, 13),
+]);
+
+/// Type-level variant of `PinMode` for alternate peripheral functions
+/// Type `C` is one of the alternate configurations
+pub struct Alternate<C: AlternateConfig> {
+    cfg: PhantomData<C>,
+}
+
+impl<C: AlternateConfig> Sealed for Alternate<C> {}
+
+//==============================================================================
+//  Pin modes
+//==============================================================================
+
+/// Type-level `enum` representing pin modes
+pub trait PinMode: Sealed + Sized {
+    /// Corresponding PinModeV
+    const MODE: PinModeV;
+    /// Value of the DIR field in this mode
+    const DIR: bool = false;
+    /// Value of the INEN field in this mode
+    const INEN: bool = false;
+    /// Value of the PULLEN field in this mode
+    const PULLEN: bool = false;
+    /// Value of the OUT field in this mode
+    const OUT: bool = false;
+    /// Value of the PMUXEN field in this mode
+    const PMUXEN: bool = false;
+    /// Value of the PMUXE/PMUXO field in this mode
+    const PMUX: u8 = 0;
+    /// Consume a `PinT` and return a new `PinT` in the requested mode
+    #[inline]
+    fn into_mode<I, M>(_pin: PinT<I, M>) -> PinT<I, Self>
+    where
+        I: PinId,
+        M: PinMode,
+    {
+        Self::convert(I::Group::group(), I::NUM);
+        PinT::new()
+    }
+    /// Convert a pin into `PinMode` `Self`.
+    ///
+    /// Users should not be able to call this function, because they
+    /// cannot access the `Group::group()` method in a private module.
+    #[inline]
+    fn convert(group: &GROUP, num: u8) {
```

Comment:
Yes, I think `#[doc(hidden)]` is a good idea. If my invariants hold, users couldn't even hope to call it though.

I don't think it should be marked `unsafe`. I don't believe there's any memory unsafety. This function only edits the registers and fields of the specific pin. And each update of those registers/fields is atomic, either through the `SET/CLR` registers, or using `WRCONFIG`. And we can guarantee that each pin is a singleton unless users call `unsafe` code.

FYI, there are some copy/paste bugs here. Notice that `PULLEN` is used for multiple incorrect fields. I just caught it tonight. I will push an update.

Reply:
Maybe the body of the function should be wrapped in `unsafe`, to communicate that the body must satisfy the invariants? I don't think the function itself needs to be unsafe though.

Right now, the function makes an implied contract that it will only modify the registers and fields of the specified pin. But I don't document that explicitly. Adding an internal `unsafe` block would communicate that and also allow the function to remain safe, because the contract is safe.

Reply:
I guess if we're writing only to bits "owned" by the given `GROUP`, we're probably OK. This takes a shared reference, so I guess in theory, if we had multiple threads and they were both calling this method on the same `GROUP` (which I guess can't happen right now, but in theory, given no constraints aside from the method's signature, might be possible), that might cause issues.

Reply:
I guess the actual write happens atomically even though the bits are being "set" in separate calls, so the interactions with `WRCONFIG` are probably fine. The rest of this is probably fine too, except perhaps for the (very) theoretical multi-threading issue I noted in my other comment.

Rather than wrapping the whole thing in an `unsafe` block, if you decide that the threading issue is something to actually worry about, just adding a "safety" section to the doc comment would probably be better than expanding the `unsafe` blocks. Keeping the size of `unsafe` blocks to a minimum is considered good practice.

Reply:
This gets at the question I pose in the main discussion thread. I agree that creating a `&GROUP` reference should be unsafe. I'm less sure that every use of it should also be unsafe. If there were a bug in the `convert` function, it could cause memory unsafety. But the origin of that unsafety is really the creation of the `&GROUP` reference in the first place.